Google.Solutions.CloudIap.IapClient/IapClient.cs

```diff
@@ -0,0 +1,177 @@
+﻿//
+// Copyright 2019 Google LLC
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Util.Store;
+using Google.Solutions.Compute;
+using Google.Solutions.Compute.Iap;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Solutions.CloudIap.IapClient
+{
+    internal class IapClient
+    {
+        private enum ExitCode : int
+        {
+            Success = 0,
+            InvalidParameters = 1,
+            UnhandledException = 2
+        }
+
+        private const string ProgramName = "IapClient";
+
+        public VmInstanceReference Instance { get; }
+        public ushort Port { get; }
+
+        public IapClient(VmInstanceReference instance, ushort port)
+        {
+            this.Instance = instance;
+            this.Port = port;
+        }
+
+        public async Task RunAsync(CancellationToken token)
+        {
+            ICredential credential = await GoogleWebAuthorizationBroker.AuthorizeAsync(
+                new ClientSecrets()
+                {
+                    ClientId = "78381520511-4fu6ve6b49kknk3dkdnpudoi0tivq6jn.apps.googleusercontent.com",
+                    ClientSecret = "dRgZl1efp_JKcUqQusuaVIrS"
+                },
+                new[] { IapTunnelingEndpoint.RequiredScope },
+                Environment.UserName,
```

Comment:
what happens if i have multple users i'd like to login to different systems.

eg.  user1@domain1  access windows VM in projectA
 i also would like to assume user2@domain2. to access windows VM in projectB ...

were we going to overrwirte the credential store here (i.,e since we're saving one file)

Reply:
Yes, that would require signing off and on again. 

I am not sure if using multiple users is really a common use case -- for now, I'd tend to say that it is probably okay to ignore that?

In the plugin, we should definitely provide a proper "Sign out" function though.

Reply:
given that you may end up using `gcloud` credentials again, a user that wants to switch could start a session, then use `gcloud config set account ...` then start another session?   that should work..one thing to try out later on is what happens whn you do switch accounts does the iap client rquest a new access token (wich woudln't be available via gcloud default credentail at that point since you swapped contexts)